src/index.js

```diff
@@ -3,52 +3,55 @@ import React from 'react'
 const changedArray = (a = [], b = []) =>
   a.length !== b.length || a.some((item, index) => !Object.is(item, b[index]))
 
-const initialState = {error: null, info: null}
+const initialState = {error: null}
 class ErrorBoundary extends React.Component {
   static getDerivedStateFromError(error) {
     return {error}
   }
 
   state = initialState
+  updatedWithError = false
   resetErrorBoundary = (...args) => {
     this.props.onReset?.(...args)
+    this.reset()
+  }
+
+  reset() {
+    this.updatedWithError = false
     this.setState(initialState)
   }
 
   componentDidCatch(error, info) {
-    this.props.onError?.(error, info?.componentStack)
-    this.setState({info})
+    this.props.onError?.(error, info)
   }
 
   componentDidUpdate(prevProps) {
-    const {error, info} = this.state
+    const {error} = this.state
     const {resetKeys} = this.props
-    if (
-      error !== null &&
-      info !== null &&
-      changedArray(prevProps.resetKeys, resetKeys)
-    ) {
+
+    // There's an edge case where if the thing that triggered the error
+    // happens to *also* be in the resetKeys array, we'd end up resetting
+    // the error boundary immediately. This would likely trigger a second
+    // error to be thrown.
+    // So we make sure that we don't check the resetKeys on the first call
+    // of cDU after the error is set
+    if (error !== null && !this.updatedWithError) {
+      this.updatedWithError = true
+      return
+    }
+
+    if (error !== null && changedArray(prevProps.resetKeys, resetKeys)) {
       this.props.onResetKeysChange?.(prevProps.resetKeys, resetKeys)
-      this.setState(initialState)
+      this.reset()
```

Comment:
Shouldn't this be calling `resetErrorBoundary`? Otherwise you won't call `props.onReset`.

Reply:
`props.onReset` will be called only for when `this.resetErrorBoundary` is called (presumably the end-user clicked a button you rendered in your fallback that triggered this). `props.onResetKeysChange` will be called only for when the reset happened due to a change in the `resetKeys`. It's possible these will both be set to the same function, but I think having the distinction could be helpful which is why they're different functions.

I'm open to your feedback on this though. Great opportunity to break things since we'll be pushing a major version bump.

Reply:
Hm. I see what you mean, but I would have assumed (if I were using this component) that `onReset` was called *any time the boundary reset* (for any reason) and `onResetKeysChange` was called *any time the reset keys changed* and those things are decoupled (because one can happen without the other).

How do you actually observe when a reset happens because of a key change? (Using `onResetKeysChange` won't be sufficient to tell you this, will it)?

Reply:
> How do you actually observe

Is "you" referring to the developer user or library author? 😅

For the developer, they'll know because `onResetKeysChange` will be called. The lib author knows because we do this check in cDU to compare the previous `resetKeys` with the next ones.

Reply:
I was referring to the library user.

And, right. I forgot that we only check and call `onResetKeysChange` if the `error !== null` (because intuitively, I still think keys changing is separate from resetting).

Anyway, maybe this isn't a valuable conversation to continue. It just seemed a little unintuitive to me, but I'm only one person.

Reply:
It's quite possible this particular feature is a bit niche. But it's definitely solved real problems for me so that's why I think it's worth the added complexity 😅

Thanks for your feedback @bvaughn! Always appreciate your time :)